xds/src/main/java/io/grpc/xds/LeastRequestLoadBalancer.java

```diff
@@ -0,0 +1,421 @@
+      Preconditions.checkArgument(!list.isEmpty(), "empty list");
+      this.list = list;
+      this.choiceCount = choiceCount;
+      this.random = random;
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      final Subchannel subchannel = nextSubchannel();
+      final OutstandingRequestsTracingFactory factory =
+          new OutstandingRequestsTracingFactory(getInFlights(subchannel));
+      return PickResult.withSubchannel(subchannel, factory);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(ReadyPicker.class).add("list", list).toString();
+    }
+
+    private Subchannel nextSubchannel() {
+      Subchannel candidate = list.get(random.nextInt(list.size()));
+      for (int i = 0; i < choiceCount - 1; ++i) {
+        Subchannel sampled = list.get(random.nextInt(list.size()));
+        if (getInFlights(sampled).get() < getInFlights(candidate).get()) {
+          candidate = sampled;
+        }
+      }
+      return candidate;
+    }
+
+    @VisibleForTesting
+    List<Subchannel> getList() {
+      return list;
+    }
+
+    @Override
+    boolean isEquivalentTo(LeastRequestPicker picker) {
+      if (!(picker instanceof ReadyPicker)) {
+        return false;
+      }
+      ReadyPicker other = (ReadyPicker) picker;
+      // the lists cannot contain duplicate subchannels
+      return other == this
+          || (list.size() == other.list.size() && new HashSet<>(list).containsAll(other.list));
+    }
+  }
+
+  @VisibleForTesting
+  static final class EmptyPicker extends LeastRequestPicker {
+
+    private final Status status;
+
+    EmptyPicker(@Nonnull Status status) {
+      this.status = Preconditions.checkNotNull(status, "status");
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      return status.isOk() ? PickResult.withNoResult() : PickResult.withError(status);
+    }
+
+    @Override
+    boolean isEquivalentTo(LeastRequestPicker picker) {
+      return picker instanceof EmptyPicker && (Objects.equal(status, ((EmptyPicker) picker).status)
+          || (status.isOk() && ((EmptyPicker) picker).status.isOk()));
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(EmptyPicker.class).add("status", status).toString();
+    }
+  }
+
+  /**
+   * A lighter weight Reference than AtomicReference.
+   */
+  static final class Ref<T> {
+    T value;
+
+    Ref(T value) {
+      this.value = value;
+    }
+  }
+
+  private static final class OutstandingRequestsTracingFactory extends
+      ClientStreamTracer.Factory {
+    private final AtomicInteger inFlights;
+
+    private OutstandingRequestsTracingFactory(AtomicInteger inFlights) {
+      this.inFlights = checkNotNull(inFlights, "inFlights");
+    }
+
+    @Override
+    public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata headers) {
+      inFlights.incrementAndGet();
```

Comment:
It might be more accurate to increment it inside `ClientStreamTracer.streamCreated()`.

Reply:
Makes sense since they are called at pretty much the same time anyway